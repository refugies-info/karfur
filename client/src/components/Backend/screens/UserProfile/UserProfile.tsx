import { useEffect, useState, useMemo } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useAsyncFn } from "react-use";
import { useTranslation } from "next-i18next";
import { StructureMemberRole, UpdateUserRequest } from "@refugies-info/api-types";
import { Col, Row } from "reactstrap";
import Button from "@codegouvfr/react-dsfr/Button";
import Input from "@codegouvfr/react-dsfr/Input";
import { Badge } from "@codegouvfr/react-dsfr/Badge";
import { ToggleSwitch } from "@codegouvfr/react-dsfr/ToggleSwitch";
import PasswordInput from "@codegouvfr/react-dsfr/blocks/PasswordInput";
import { logger } from "logger";
import API from "utils/API";
import { setAuthToken } from "utils/authToken";
import { isValidEmail, isValidPhone } from "lib/validateFields";
import { cls } from "lib/classname";
import { getPasswordStrength } from "lib/validatePassword";
import { userDetailsSelector, userSelector } from "services/User/user.selectors";
import { fetchUserActionCreator } from "services/User/user.actions";
import { userStructureRoleSelector, userStructureSelector } from "services/UserStructure/userStructure.selectors";
import {
  EditAvatar,
  EditButton,
  LanguageBadge,
  ModalDepartments,
  modalDepartments,
  ModalLanguage,
  modalLanguage,
  modalResetPassword,
  ModalResetPassword,
  Tag,
} from "./components";
import ErrorMessage from "components/UI/ErrorMessage";
import FRLink from "components/UI/FRLink";
import styles from "./UserProfile.module.scss";

interface Props {
  title: string;
}

export const UserProfile = (props: Props) => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  const user = useSelector(userSelector);
  const userDetails = useSelector(userDetailsSelector);
  const userStructure = useSelector(userStructureSelector);
  const userStructureRole = useSelector(userStructureRoleSelector);

  const [edition, setEdition] = useState(false);
  const [username, setUsername] = useState<string>(userDetails?.username || "");
  const [usernameError, setUsernameError] = useState<string | null>(null);
  const [firstName, setFirstName] = useState<string>(userDetails?.firstName || "");
  const [email, setEmail] = useState<string>(userDetails?.email || "");
  const [emailHint, setEmailHint] = useState<boolean>(false);
  const [emailError, setEmailError] = useState<string | null>(null);
  const [phone, setPhone] = useState<string>(userDetails?.phone || "");
  const [phoneError, setPhoneError] = useState<string | null>(null);
  const [newsletter, setNewsletter] = useState<boolean | null>(null);
  const [nbWordsTranslated, setNbWordsTranslated] = useState<number | null>(null);
  const [oldPassword, setOldPassword] = useState("");
  const [newPassword, setNewPassword] = useState("");
  const [passwordError, setPasswordError] = useState<string | null>(null);
  const passwordStrength = useMemo(() => getPasswordStrength(newPassword), [newPassword]);

  useEffect(() => {
    document.title = props.title;
  }, [props.title]);

  useEffect(() => {
    if (!userDetails) return;
    API.isInContacts().then((res) => setNewsletter(res.isInContacts));
  }, [dispatch, user, userDetails, nbWordsTranslated]);

  const subscribeNewsletter = async (checked: boolean) => {
    if (!userDetails?.email) return;
    setNewsletter(checked);
    try {
      if (checked) {
        await API.contacts({ email: userDetails.email });
      } else {
        await API.deleteContact();
      }
    } catch (e: any) {
      logger.error(e);
    }
  };

  // Load user stats
  useEffect(() => {
    if (!userDetails) return;
    const loadIndicators = async () => {
      if (user.traducteur || user.expertTrad) {
        const data = await API.get_progression({ onlyTotal: true });
        setNbWordsTranslated(data?.totalIndicator?.wordsCount || 0);
      }
    };
    if (nbWordsTranslated === null) {
      loadIndicators();
    }
  }, [dispatch, user, userDetails, nbWordsTranslated]);

  // Fill form
  useEffect(() => {
    if (!userDetails) return;
    if (userDetails.username) setUsername(userDetails.username);
    if (userDetails.firstName) setFirstName(userDetails.firstName);
    if (userDetails.email) setEmail(userDetails.email);
    if (userDetails.phone) setPhone(userDetails.phone);
  }, [dispatch, userDetails]);

  // Form validation
  useEffect(() => {
    if (phone && !isValidPhone(phone))
      setPhoneError("Ce n'est pas un numéro de téléphone valide, vérifiez votre saisie.");
    else setPhoneError(null);
  }, [phone]);

  useEffect(() => {
    if (!email) setEmailError("L'email est obligatoire");
    else if (!isValidEmail(email)) setEmailError("Ce n'est pas une adresse email valide, vérifiez votre saisie.");
    else setEmailError(null);
  }, [email]);

  useEffect(() => {
    const usernameMandatory = user.admin || user.contributeur || user.expertTrad || user.traducteur;
    if (!username && usernameMandatory) setUsernameError("Veuillez choisir un pseudonyme.");
    else setUsernameError(null); // needed to reset api errors
  }, [username, user]);

  const [{ loading, error }, submit] = useAsyncFn(
    async (e: any) => {
      e.preventDefault();
      setPasswordError(null);
      if (!userDetails || phoneError || emailError || usernameError) return;
      const updateUserRequest: UpdateUserRequest = {
        action: "modify-my-details",
        user: {
          username,
          firstName,
          email,
          phone,
        },
      };
      if (!oldPassword && newPassword) {
        setPasswordError("Veuillez renseigner votre ancien mot de passe.");
        return;
      }
      if (oldPassword && newPassword && !userDetails.sso) {
        updateUserRequest.user.password = {
          oldPassword,
          newPassword,
        };
      }
      try {
        const data = await API.updateUser(userDetails._id, updateUserRequest);
        if (data.token) setAuthToken(data.token);
        dispatch(fetchUserActionCreator());
        setEdition(false);
        // reset password inputs
        if (updateUserRequest.user.password) {
          setOldPassword("");
          setNewPassword("");
        }
      } catch (e: any) {
        const errorCode = e.response?.data?.code;
        if (errorCode === "NO_CODE_SUPPLIED") {
          // TODO: handle 2FA
        } else if (errorCode === "INVALID_PASSWORD") {
          setPasswordError("Mot de passe incorrect, vérifiez votre saisie");
        } else if (errorCode === "USED_PASSWORD") {
          setPasswordError("Le nouveau mot de passe doit être différent du mot de passe actuel actuel.");
        } else if (errorCode === "PASSWORD_TOO_WEAK") {
          setPasswordError(
            "Votre mot de passe ne contient pas l'un des éléments suivants : 7 caractères minimum, 1 caractère spécial, 1 chiffre minimum.",
          );
        } else if (errorCode === "USERNAME_TAKEN") {
          setUsernameError("Ce nom d'utilisateur est déjà pris. Veuillez en choisir un autre.");
        } else if (errorCode === "EMAIL_TAKEN") {
          setEmailError("Un compte Réfugiés.info est déjà associé à cette adresse email.");
        } else {
          throw new Error("Une erreur est survenue. Veuillez réessayer");
        }
      }
    },
    [userDetails, username, firstName, email, phone, phoneError, emailError, usernameError, oldPassword, newPassword],
  );

  if (!userDetails) return <div>Une erreur est survenue, veuillez recharger la page&nbsp;!</div>;

  return (
    <div className={styles.container}>
      <div className={styles.wrapper}>
        <Row className="mb-10 align-items-center">
          <Col>
            <h1>Mon profil</h1>
          </Col>
          <Col className="text-end text-nowrap">
            {/* <Button
              disabled={edition}
              priority="secondary"
              size="small"
              iconId="fr-icon-error-warning-line"
              iconPosition="right"
              className={cls(styles.danger, "ms-2")}
            >
              Supprimer mon profil
            </Button> */}
          </Col>
        </Row>

        <Row>
          <Col className="flex-grow-0">
            <h2>Photo de profil</h2>
            <div className={styles.block}>
              <EditAvatar />
            </div>

            <h2 className="my-6">Activité</h2>
            {nbWordsTranslated !== null && (
              <div className={styles.info}>
                <label>Nombre de mots traduits</label>
                <p>{nbWordsTranslated}</p>
              </div>
            )}
            {userStructure && (
              <div className={styles.info}>
                <p>{userStructure.nom}</p>
                {userStructureRole?.includes(StructureMemberRole.ADMIN) && (
                  <Badge severity="success" noIcon>
                    Responsable
                  </Badge>
                )}
              </div>
            )}

            {user.caregiver && userDetails.partner && (
              <div className={styles.info}>
                <label>Partenaire</label>
                <p>{userDetails.partner}</p>
              </div>
            )}

            {(user.admin || user.expertTrad) && (
              <div className={styles.info}>
                <label className="mb-2">Rôles exceptionnels</label>
                {user.admin && <Tag className="w-100 mb-2">Administrateur</Tag>}
                {user.expertTrad && <Tag className="w-100">Expert en traduction</Tag>}
              </div>
            )}
          </Col>

          <Col>
            <div className="d-flex align-content-center justify-content-between">
              <h2>Informations personnelles</h2>
              {edition ? (
                <Button
                  priority="primary"
                  size="small"
                  onClick={submit}
                  iconId="fr-icon-save-3-line"
                  iconPosition="right"
                  disabled={loading}
                >
                  Sauvegarder les modifications
                </Button>
              ) : (
                <Button
                  priority="secondary"
                  size="small"
                  onClick={() => setEdition(true)}
                  iconId="fr-icon-edit-box-line"
                  iconPosition="right"
                >
                  Modifier mon profil
                </Button>
              )}
            </div>

            <div className={cls(styles.block, styles.form, "mb-10")}>
              <form onSubmit={submit}>
                <Input
                  label="Pseudonyme public"
                  state={!!usernameError && edition ? "error" : "default"}
                  stateRelatedMessage={usernameError}
                  hintText="Votre pseudonyme apparaît sur les fiches auxquelles vous contribuez."
                  nativeInputProps={{
                    name: "pseudo",
                    readOnly: !edition,
                    value: username || (!edition ? "Non défini" : ""),
                    onChange: (e: any) => setUsername(e.target.value),
                    title: "pseudo-input",
                  }}
                  disabled={loading}
                  className={!username ? styles.empty : ""}
                />
                <Input
                  label="Prénom"
                  hintText="Votre prénom n’est pas public, il est utilisé pour échanger avec vous directement."
                  nativeInputProps={{
                    name: "firstName",
                    readOnly: !edition,
                    value: firstName || (!edition ? "Non défini" : ""),
                    onChange: (e: any) => setFirstName(e.target.value),
                    title: "firstname-input",
                  }}
                  disabled={loading}
                  className={!firstName ? styles.empty : ""}
                />
                <Input
                  label="Email"
                  state={!!emailError && edition ? "error" : "default"}
                  stateRelatedMessage={emailError}
                  hintText={
                    emailHint
                      ? "Saisissez la nouvelle adresse email que vous souhaitez associer à votre compte. Nous enverrons un code de vérification à cette adresse."
                      : null
                  }
                  nativeInputProps={{
                    name: "email",
                    readOnly: !edition,
                    value: email || (!edition ? "Non défini" : ""),
                    onChange: (e: any) => setEmail(e.target.value),
                    onFocus: () => setEmailHint(true),
                    onBlur: () => setEmailHint(false),
                    title: "email-input",
                  }}
                  disabled={loading}
                  className={!email ? styles.empty : ""}
                />
                <Input
                  label="Téléphone"
                  state={!!phoneError && edition ? "error" : "default"}
                  stateRelatedMessage={phoneError}
                  nativeInputProps={{
                    name: "phone",
                    readOnly: !edition,
                    value: phone || (!edition ? "Non défini" : ""),
                    onChange: (e: any) => setPhone(e.target.value),
                    title: "phone-input",
                  }}
                  disabled={loading}
                  className={!phone ? styles.empty : ""}
                />
                {edition && !userDetails.sso && (
                  <>
                    <PasswordInput
                      label="Mot de passe actuel"
                      messages={[]}
                      nativeInputProps={{
                        name: "old-password",
                        value: oldPassword,
                        onChange: (e: any) => setOldPassword(e.target.value),
                        title: "old-password-input",
                      }}
                    />
                    <FRLink onClick={() => modalResetPassword.open()} className={styles.link}>
                      Mot de passe oublié&nbsp;?
                    </FRLink>
                    <PasswordInput
                      label="Nouveau mot de passe"
                      messages={passwordStrength.criterias.map((criteria) => ({
                        message: t(criteria.label),
                        severity: !newPassword ? "info" : criteria.isOk ? "valid" : "error",
                      }))}
                      nativeInputProps={{
                        name: "new-password",
                        value: newPassword,
                        onChange: (e: any) => setNewPassword(e.target.value),
                        title: "new-password-input",
                      }}
                      className="mt-3"
                    />
                    <ErrorMessage error={passwordError} />
                  </>
                )}
                <ErrorMessage error={error?.message} />
              </form>
            </div>

            <h2>Préférences</h2>
            <div className={cls(styles.block, "mb-4")}>
              <div className="d-flex justify-content-between mb-3">
                <label className={styles.label}>Départements pour la recherche</label>
                <EditButton icon="map" onClick={() => modalDepartments.open()} />
              </div>
              {!userDetails.departments ? (
                <p className={styles.empty}>Non défini</p>
              ) : (
                userDetails.departments.map((dep, i) => (
                  <Tag key={i} className="me-1 mt-1">
                    {dep}
                  </Tag>
                ))
              )}
            </div>

            {(user.traducteur || user.expertTrad) && (
              <div className={cls(styles.block, "mb-4")}>
                <div className="d-flex justify-content-between mb-4">
                  <label className={styles.label}>Langues de traduction</label>
                  <EditButton icon="translate" onClick={() => modalLanguage.open()} />
                </div>
                {!userDetails.selectedLanguages ? (
                  <p className={styles.empty}>Non défini</p>
                ) : (
                  userDetails.selectedLanguages.map((ln, i) => <LanguageBadge key={i} id={ln} />)
                )}
              </div>
            )}

            <div className={styles.block}>
              <label className={cls(styles.label, "mb-2")} htmlFor="newsletter">
                Communication par email
              </label>

              <ToggleSwitch
                label={
                  <span>
                    <strong>La lettre d’information de Réfugiés.info</strong> : actualités, nouveaux contenus, mises à
                    jour, événements.
                  </span>
                }
                helperText="Maximum 1 fois par mois"
                checked={!!newsletter}
                onChange={subscribeNewsletter}
                disabled={newsletter === null}
                showCheckedHint={false}
              />
            </div>
          </Col>
        </Row>
      </div>

      <ModalDepartments />
      <ModalLanguage />
      <ModalResetPassword email={userDetails?.email || ""} />
    </div>
  );
};

export default UserProfile;
