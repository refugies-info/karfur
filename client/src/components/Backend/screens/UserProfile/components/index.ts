export { default as LanguageBadge } from "./LanguageBadge";
export { default as Tag } from "./Tag";
export { default as EditButton } from "./EditButton";
export { default as EditAvatar } from "./EditAvatar";
export * from "./ModalDepartments";
export * from "./ModalLanguage";
export * from "./ModalResetPassword";
export * from "./ModalEmailCode";
