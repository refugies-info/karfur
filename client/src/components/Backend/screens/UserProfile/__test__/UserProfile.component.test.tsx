import { UserProfile } from "../UserProfile";
import { act, ReactTestRenderer } from "react-test-renderer";
import {
  wrapWithProvidersAndRender,
  wrapWithProvidersAndRenderForTesting,
} from "../../../../../../jest/lib/wrapWithProvidersAndRender";
import { initialMockStore } from "__fixtures__/reduxStore";
import { testUser } from "__fixtures__/user";
import { fireEvent, RenderResult, waitFor } from "@testing-library/react";
import "jest-styled-components";
import { setupGoogleMock } from "__mocks__/react-google-autocomplete";
import "@testing-library/jest-dom";

jest.mock("next/router", () => require("next-router-mock"));
jest.mock("components/UI/Tooltip", () => jest.fn().mockReturnValue(<></>));
jest.mock("@codegouvfr/react-dsfr/SearchBar", () => ({ SearchBar: jest.fn().mockReturnValue(<></>) }));

jest.mock("utils/API", () => ({
  __esModule: true, // this property makes it work
  default: {
    updateUser: jest.fn(),
    isInContacts: jest.fn().mockResolvedValue({ isInContacts: false }),
    updatePassword: jest.fn(),
  },
}));

jest.mock("services/User/user.actions", () => {
  const actions = jest.requireActual("services/User/user.actions");
  return {
    saveUserActionCreator: jest.fn(actions.saveUserActionCreator),
    fetchUserActionCreator: jest.fn(actions.fetchUserActionCreator),
  };
});

describe("UserProfile", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupGoogleMock();
  });

  it("should render correctly when loading", () => {
    window.scrollTo = jest.fn();
    let component: ReactTestRenderer;
    component = wrapWithProvidersAndRender({
      Component: UserProfile,
      reduxState: {
        ...initialMockStore,
        loadingStatus: { FETCH_USER: { isLoading: true } },
      },
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render correctly", () => {
    window.scrollTo = jest.fn();
    let component: ReactTestRenderer;
    component = wrapWithProvidersAndRender({
      Component: UserProfile,
      reduxState: {
        ...initialMockStore,
        user: { ...initialMockStore.user, user: { ...testUser } },
      },
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render correctly when editing profile", async () => {
    window.scrollTo = jest.fn();
    let component: RenderResult;
    act(() => {
      component = wrapWithProvidersAndRenderForTesting({
        Component: UserProfile,
        reduxState: {
          ...initialMockStore,
          user: { ...initialMockStore.user, user: { ...testUser } },
        },
      });
    });
    act(() => {
      fireEvent.click(component.getByText("Modifier mon profil"));
    });
    await waitFor(() => expect(component.getAllByRole("textbox")[0]).not.toBeDisabled());
  });
});
