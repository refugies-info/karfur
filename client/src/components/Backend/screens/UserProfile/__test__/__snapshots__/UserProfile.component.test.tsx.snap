// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="wrapper"
  >
    <div
      className="mb-10 align-items-center row"
    >
      <div
        className="col"
      >
        <h1>
          Mon profil
        </h1>
      </div>
      <div
        className="text-end text-nowrap col"
      >
        <button
          className="fr-btn fr-btn--secondary fr-btn--sm fr-icon-error-warning-line fr-btn--icon-right danger ms-2"
          disabled={false}
          id="fr-button-:r0:"
          onClick={[Function]}
          style={{}}
        >
          Supprimer mon profil
        </button>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="flex-grow-0 col"
      >
        <h2>
          Photo de profil
        </h2>
        <div
          className="block"
        >
          <div
            className="container"
          >
            <div
              className="avatar"
            >
              <img
                alt="user picture"
                data-nimg="1"
                decoding="async"
                height={160}
                loading="lazy"
                onError={[Function]}
                onLoad={[Function]}
                src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
                srcSet="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
                style={
                  {
                    "color": "transparent",
                  }
                }
                width={160}
              />
            </div>
            <p
              className="small my-4"
            >
              Votre photo apparaît sur les fiches auxquelles vous contribuez.
            </p>
            <button
              className="fr-btn fr-btn--secondary fr-btn--sm fr-icon-image-edit-line fr-btn--icon-right position-relative"
            >
              Modifier ma photo
              <input
                id="avatar"
                onChange={[Function]}
                type="file"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="d-flex align-items-center justify-content-between"
        >
          <h2>
            Informations personnelles
          </h2>
          <button
            className="fr-btn fr-btn--secondary fr-btn--sm fr-icon-edit-box-line fr-btn--icon-right fix_align"
            id="fr-button-:r1:"
            onClick={[Function]}
            style={{}}
          >
            Modifier mes informations personnelles
          </button>
        </div>
        <div
          className="block form mb-10"
        >
          <form
            onSubmit={[Function]}
          >
            <div
              className="fr-input-group"
            >
              <label
                className="fr-label"
                htmlFor="input-:r2:"
              >
                Pseudonyme public
                <span
                  className="fr-hint-text"
                >
                  Votre pseudonyme apparaît sur les fiches auxquelles vous contribuez.
                </span>
              </label>
              <input
                aria-describedby="input-:r2:-desc-error"
                className="fr-input"
                id="input-:r2:"
                name="pseudo"
                onChange={[Function]}
                readOnly={true}
                title="pseudo-input"
                type="text"
                value="test"
              />
            </div>
            <div
              className="fr-input-group empty"
            >
              <label
                className="fr-label"
                htmlFor="input-:r3:"
              >
                Prénom
                <span
                  className="fr-hint-text"
                >
                  Votre prénom n’est pas public, il est utilisé pour échanger avec vous directement.
                </span>
              </label>
              <input
                aria-describedby="input-:r3:-desc-error"
                className="fr-input"
                id="input-:r3:"
                name="firstName"
                onChange={[Function]}
                readOnly={true}
                title="firstname-input"
                type="text"
                value="Non défini"
              />
            </div>
            <div
              className="fr-input-group empty"
            >
              <label
                className="fr-label"
                htmlFor="input-:r4:"
              >
                Email
                <span
                  className="fr-hint-text"
                />
              </label>
              <input
                aria-describedby="input-:r4:-desc-error"
                className="fr-input"
                id="input-:r4:"
                name="email"
                onChange={[Function]}
                readOnly={true}
                title="email-input"
                type="text"
                value="Non défini"
              />
            </div>
            <div
              className="fr-input-group empty"
            >
              <label
                className="fr-label"
                htmlFor="input-:r5:"
              >
                Téléphone
              </label>
              <input
                aria-describedby="input-:r5:-desc-error"
                className="fr-input"
                id="input-:r5:"
                name="phone"
                onChange={[Function]}
                readOnly={true}
                title="phone-input"
                type="text"
                value="Non défini"
              />
            </div>
            <div
              className="fr-input-group"
            >
              <label
                className="fr-label"
                htmlFor="input-:r6:"
              >
                Mot de passe actuel
              </label>
              <input
                aria-describedby="input-:r6:-desc-error"
                className="fr-input"
                id="input-:r6:"
                name="fake-password-input"
                readOnly={true}
                title="fake-password-input"
                type="text"
                value="********"
              />
            </div>
            <button
              className="fr-link link link"
              onClick={[Function]}
              type="button"
            >
              Mot de passe oublié ?
            </button>
          </form>
          <button
            aria-controls="user-email-code-modal"
            aria-hidden={true}
            className="fr-btn fr-hidden"
            data-fr-opened={false}
            id="user-email-code-modal-hidden-control-button"
            style={{}}
            tabIndex={-1}
            type="button"
          >
             
          </button>
          <dialog
            aria-labelledby="fr-modal-title-user-email-code-modal"
            className="fr-modal"
            data-fr-concealing-backdrop={true}
            id="user-email-code-modal"
            role="dialog"
          >
            <div
              className="fr-container fr-container--fluid fr-container-md"
            >
              <div
                className="fr-grid-row fr-grid-row--center"
              >
                <div
                  className="fr-col-12 fr-col-md-8 fr-col-lg-6"
                >
                  <div
                    className="fr-modal__body"
                  >
                    <div
                      className="fr-modal__header"
                    >
                      <button
                        aria-controls="user-email-code-modal"
                        className="fr-btn--close fr-btn"
                        title="Fermer"
                        type="button"
                      >
                        Fermer
                      </button>
                    </div>
                    <div
                      className="fr-modal__content"
                    >
                      <h1
                        className="fr-modal__title"
                        id="fr-modal-title-user-email-code-modal"
                      >
                        Entrez le code reçu
                      </h1>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </dialog>
          <button
            aria-controls="user-password-modal"
            aria-hidden={true}
            className="fr-btn fr-hidden"
            data-fr-opened={false}
            id="user-password-modal-hidden-control-button"
            style={{}}
            tabIndex={-1}
            type="button"
          >
             
          </button>
          <dialog
            aria-labelledby="fr-modal-title-user-password-modal"
            className="fr-modal"
            data-fr-concealing-backdrop={true}
            id="user-password-modal"
            role="dialog"
          >
            <div
              className="fr-container fr-container--fluid fr-container-md"
            >
              <div
                className="fr-grid-row fr-grid-row--center"
              >
                <div
                  className="fr-col-12 fr-col-md-8 fr-col-lg-6"
                >
                  <div
                    className="fr-modal__body"
                  >
                    <div
                      className="fr-modal__header"
                    >
                      <button
                        aria-controls="user-password-modal"
                        className="fr-btn--close fr-btn"
                        title="Fermer"
                        type="button"
                      >
                        Fermer
                      </button>
                    </div>
                    <div
                      className="fr-modal__content"
                    >
                      <h1
                        className="fr-modal__title"
                        id="fr-modal-title-user-password-modal"
                      >
                        Mot de passe oublié
                      </h1>
                      <p>
                        Nous allons vous envoyer un mail avec des instructions pour le réinitialiser.
                      </p>
                      <form
                        className="container"
                        onSubmit={[Function]}
                      >
                        <div
                          className="fr-input-group mb-0"
                        >
                          <label
                            className="fr-label"
                            htmlFor="input-:r9:"
                          >
                            Adresse mail
                          </label>
                          <input
                            aria-describedby="input-:r9:-desc-error"
                            autoFocus={true}
                            className="fr-input"
                            defaultValue=""
                            id="input-:r9:"
                            name="email"
                            type="email"
                          />
                        </div>
                        <div
                          className="text-end"
                        >
                          <button
                            className="fr-btn fr-icon-mail-line fr-btn--icon-right button mt"
                            id="fr-button-:ra:"
                            style={{}}
                            type="submit"
                          >
                            Envoyer le lien de réinitialisation
                          </button>
                        </div>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </dialog>
        </div>
        <h2>
          Préférences
        </h2>
        <div
          className="block mb-4"
        >
          <div
            className="d-flex justify-content-between mb-3"
          >
            <label
              className="label"
            >
              Départements pour la recherche
            </label>
            <button
              className="button"
              onClick={[Function]}
            >
              <i
                className="fr-icon-map-pin-2-line icon"
              />
              Modifier
            </button>
          </div>
          <p
            className="empty"
          >
            Non défini
          </p>
        </div>
        <div
          className="block"
        >
          <label
            className="label mb-2"
            htmlFor="newsletter"
          >
            Communication par email
          </label>
          <div
            className="fr-toggle"
            id="fr-toggle-:rb:"
          >
            <input
              aria-describedby="fr-toggle-:rb:-hint-text"
              checked={false}
              className="fr-toggle__input"
              disabled={true}
              id="fr-toggle-:rb:-input"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="fr-toggle__label"
              htmlFor="fr-toggle-:rb:-input"
            >
              <span>
                <strong>
                  La lettre d’information de Réfugiés.info
                </strong>
                 : actualités, nouveaux contenus, mises à jour, événements.
              </span>
            </label>
            <p
              className="fr-hint-text"
              id="fr-toggle-:rb:-hint-text"
            >
              Maximum 1 fois par mois
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    aria-controls="user-departments-modal"
    aria-hidden={true}
    className="fr-btn fr-hidden"
    data-fr-opened={false}
    id="user-departments-modal-hidden-control-button"
    style={{}}
    tabIndex={-1}
    type="button"
  >
     
  </button>
  <dialog
    aria-labelledby="fr-modal-title-user-departments-modal"
    className="fr-modal"
    data-fr-concealing-backdrop={true}
    id="user-departments-modal"
    role="dialog"
  >
    <div
      className="fr-container fr-container--fluid fr-container-md"
    >
      <div
        className="fr-grid-row fr-grid-row--center"
      >
        <div
          className="fr-col-12 fr-col-md-8 fr-col-lg-6"
        >
          <div
            className="fr-modal__body"
          >
            <div
              className="fr-modal__header"
            >
              <button
                aria-controls="user-departments-modal"
                className="fr-btn--close fr-btn"
                title="Fermer"
                type="button"
              >
                Fermer
              </button>
            </div>
            <div
              className="fr-modal__content"
            >
              <h1
                className="fr-modal__title"
                id="fr-modal-title-user-departments-modal"
              >
                Départements pour la recherche
              </h1>
              <p
                className="mb-8"
              >
                Nous vous montrerons uniquement les contenus susceptibles de vous intéresser. Les démarches nationales resteront visibles.
              </p>
              <form
                onSubmit={[Function]}
              >
                <label
                  className="mb-2"
                  htmlFor="location"
                >
                  Nom ou numéro du département
                </label>
                <div
                  className="position-relative"
                >
                  <div />
                </div>
                <div
                  className="text-end"
                >
                  <button
                    className="fr-btn fr-icon-check-line fr-btn--icon-right button mt-12"
                    disabled={true}
                    id="fr-button-:rd:"
                    style={{}}
                    type="submit"
                  >
                    Valider
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </dialog>
  <button
    aria-controls="user-language-modal"
    aria-hidden={true}
    className="fr-btn fr-hidden"
    data-fr-opened={false}
    id="user-language-modal-hidden-control-button"
    style={{}}
    tabIndex={-1}
    type="button"
  >
     
  </button>
  <dialog
    aria-labelledby="fr-modal-title-user-language-modal"
    className="fr-modal"
    data-fr-concealing-backdrop={true}
    id="user-language-modal"
    role="dialog"
  >
    <div
      className="fr-container fr-container--fluid fr-container-md"
    >
      <div
        className="fr-grid-row fr-grid-row--center"
      >
        <div
          className="fr-col-12 fr-col-md-8 fr-col-lg-6"
        >
          <div
            className="fr-modal__body"
          >
            <div
              className="fr-modal__header"
            >
              <button
                aria-controls="user-language-modal"
                className="fr-btn--close fr-btn"
                title="Fermer"
                type="button"
              >
                Fermer
              </button>
            </div>
            <div
              className="fr-modal__content"
            >
              <h1
                className="fr-modal__title"
                id="fr-modal-title-user-language-modal"
              >
                <div
                  className="mb-6"
                >
                  Langues de traduction
                </div>
              </h1>
              <form
                onSubmit={[Function]}
              >
                <label
                  className="choice mb-4 rounded-0"
                  htmlFor="language"
                  onClick={[Function]}
                >
                  <input
                    defaultChecked={false}
                    name="language"
                    type="checkbox"
                    value="en"
                  />
                  <i
                    className="ri-checkbox-blank-line me-2"
                  />
                  <span
                    className="text"
                  >
                    <strong>
                      Anglais
                    </strong>
                     - 
                    English
                  </span>
                  <span
                    className="flag fi fi-en"
                    title="en"
                  />
                </label>
                <div
                  className="text-end"
                >
                  <button
                    className="fr-btn fr-icon-check-line fr-btn--icon-right button mt-8"
                    disabled={true}
                    id="fr-button-:rf:"
                    style={{}}
                    type="submit"
                  >
                    Valider
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </dialog>
  <button
    aria-controls="delete-account-modal"
    aria-hidden={true}
    className="fr-btn fr-hidden"
    data-fr-opened={false}
    id="delete-account-modal-hidden-control-button"
    style={{}}
    tabIndex={-1}
    type="button"
  >
     
  </button>
  <dialog
    aria-labelledby="fr-modal-title-delete-account-modal"
    className="fr-modal"
    data-fr-concealing-backdrop={true}
    id="delete-account-modal"
    role="dialog"
  >
    <div
      className="fr-container fr-container--fluid fr-container-md"
    >
      <div
        className="fr-grid-row fr-grid-row--center"
      >
        <div
          className="fr-col-12 fr-col-md-8 fr-col-lg-6"
        >
          <div
            className="fr-modal__body"
          >
            <div
              className="fr-modal__header"
            >
              <button
                aria-controls="delete-account-modal"
                className="fr-btn--close fr-btn"
                title="Fermer"
                type="button"
              >
                Fermer
              </button>
            </div>
            <div
              className="fr-modal__content"
            >
              <h1
                className="fr-modal__title"
                id="fr-modal-title-delete-account-modal"
              >
                Êtes-vous sûrs de vouloir supprimer votre compte ?
              </h1>
              <p
                className="mb-8"
              >
                Vous perdrez l'accès à vos favoris ainsi qu'à votre structure et aux fiches que vous avez rédigées.
              </p>
            </div>
            <div
              className="fr-modal__footer"
            >
              <ul
                className="fr-btns-group fr-btns-group--right fr-btns-group--inline-reverse fr-btns-group--inline-lg fr-btns-group--icon-left"
              >
                <li>
                  <button
                    aria-controls="delete-account-modal"
                    className="fr-btn fr-icon-delete-bin-line fr-btn--icon-left danger"
                    disabled={false}
                    id="fr-button-:rh:"
                    onClick={[Function]}
                    style={{}}
                  >
                    Supprimer mon compte
                  </button>
                </li>
                <li>
                  <button
                    aria-controls="delete-account-modal"
                    className="fr-btn fr-btn--secondary"
                    disabled={false}
                    id="fr-button-:ri:"
                    onClick={[Function]}
                    style={{}}
                  >
                    Annuler
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </dialog>
</div>
`;

exports[`UserProfile should render correctly when loading 1`] = `
<div>
  Une erreur est survenue, veuillez recharger la page !
</div>
`;
